tests/HopSpec/AnfSpec.hs

```diff
@@ -0,0 +1,340 @@
+
+import Test.Hspec
+import Test.Hspec.Expectations
+import qualified Data.Text as T
+import qualified Data.Vector as V
+
+spec :: Spec
+spec =
+  describe "ANF" $ do
+    describe "toAnf" $ do
+      let v0 = LocalVar $ LocalNamelessVar 0 $ BinderSlot 0
+          v0_0 = v0
+          v0_1 = LocalVar $ LocalNamelessVar 0 $ BinderSlot 1
+          v1 = LocalVar $ LocalNamelessVar 1 $ BinderSlot 0
+          v2 = LocalVar $ LocalNamelessVar 2 $ BinderSlot 0
+          add = GlobalVarSym $ GlobalSymbol $ T.pack "add"
+          abs = GlobalVarSym $ GlobalSymbol $ T.pack "abs"
+          neg = GlobalVarSym $ GlobalSymbol $ T.pack "neg"
+          id_ = GlobalVarSym $ GlobalSymbol $ T.pack "id"
+          ten = LInteger 10
+          twenty = LInteger 20
+          dummyBI = BinderInfoData Omega () Nothing
+          prim0 = PrimopIdGeneral "test"
+
+      describe "simple tail cases" $ do
+        it "converts variables" $
+          let term = V v0
+              anf = AnfReturn $ V.singleton v0
+          in toAnf term `shouldBe` anf
+
+        -- TODO: need to add pass to remove shifts
+        --
+        -- it "converts shifted variables" $
+        --   let term = BinderLevelShiftUP 2 $ V v0
+        --       anf = AnfShift 2 $ AnfReturn $ V.singleton v0
+        --   in toAnf term `shouldBe` anf
+
+        it "converts literals" $
+          let lit = LInteger 42
+              term = ELit lit
+              anf = AnfLet (Arity 1)
+                           (RhsAlloc $ AllocLit lit)
+                           (AnfReturn $ V.singleton v0)
+          in toAnf term `shouldBe` anf
+
+        it "converts returns" $
+          let term = Return $ V.fromList [ELit ten, V v0]
+              anf  = AnfLet (Arity 1)
+                            (RhsAlloc $ AllocLit ten)
+                            (AnfReturn $ V.fromList [v0, v1])
+           in toAnf term `shouldBe` anf
+
+        it "converts delays" $
+          let -- delay ((0) (0))
+              term = Delay $ App (V v0) $ V.singleton (V v0)
+              anf  = AnfLet (Arity 1)
+                            (RhsAlloc $
+                              AllocThunk $
+                               AnfTailCall $ AppFun v0 $ V.singleton v0)
+                            (AnfReturn $ V.singleton v0)
+          in toAnf term `shouldBe` anf
+
+        it "converts thunk invocations" $
+          let -- force (delay 10)
+              term = EnterThunk (Delay $ ELit ten)
+              -- letA (delay (letA 10 in (0)))
+              -- in   force (0)
+              anf  = AnfLet (Arity 1)
+                            (RhsAlloc $
+                              AllocThunk (AnfLet (Arity 1)
+                                                 (RhsAlloc $ AllocLit ten)
+                                                 (AnfReturn $ V.singleton v0)))
+                            (AnfTailCall $ AppThunk v0)
+          in toAnf term `shouldBe` anf
+
+        it "converts no-arg tail calls" $
```

Comment:
Yes.  You can have a unboxed tuple of zero arity, which is different than a thunk because it doesn't cache its answer.  

The types are calling conventions paper makes the convention that zero arity lambdas are thunks, but we distinguish between them.